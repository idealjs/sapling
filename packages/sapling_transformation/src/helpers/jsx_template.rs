use biome_js_syntax::{
    AnyJsAssignmentPattern, AnyJsBinding, AnyJsBindingPattern, AnyJsCallArgument,
    JsIdentifierBinding,
};
use biome_rowan::AstNode;
use std::any::Any;

use biome_js_factory::make::{
    js_arrow_function_expression, js_assignment_expression, js_call_argument_list,
    js_call_arguments, js_call_expression, js_computed_member_assignment, js_directive_list,
    js_expression_statement, js_function_body, js_function_expression, js_identifier_assignment,
    js_identifier_binding, js_identifier_expression, js_name, js_parameter_list, js_parameters,
    js_parenthesized_expression, js_reference_identifier, js_return_statement, js_statement_list,
    js_static_member_assignment, js_static_member_expression, js_variable_declaration,
    js_variable_declarator, js_variable_declarator_list, js_variable_statement, token,
};
use biome_js_syntax::{
    AnyJsExpression, AnyJsParameter, AnyJsStatement, AnyJsxChild, JsArrowFunctionExpression,
    JsCallExpression, JsDirectiveList, JsExpressionStatement, JsFunctionBody, JsFunctionExpression,
    JsLanguage, JsParameterList, JsParameters, JsReturnStatement, JsStatementList, JsSyntaxToken,
    JsxElement, JsxTagExpression, T,
};

use biome_rowan::SyntaxToken;

pub fn make_js_call_expression(
    arrow_function_expression: JsArrowFunctionExpression,
) -> JsCallExpression {
    let callee = js_parenthesized_expression(
        token(T!['(']),
        AnyJsExpression::JsArrowFunctionExpression(arrow_function_expression),
        token(T![')']),
    );

    let arguments = js_call_arguments(
        token(T!['(']),
        js_call_argument_list([], []),
        token(T![')']),
    );

    let expression = js_call_expression(
        AnyJsExpression::JsParenthesizedExpression(callee),
        arguments,
    )
    .build();

    expression
}

pub fn make_js_arrow_function_expression(
    params: JsParameters,
    body: JsFunctionBody,
) -> JsArrowFunctionExpression {
    use biome_js_syntax::{AnyJsArrowFunctionParameters, AnyJsFunctionBody, T};

    let params = AnyJsArrowFunctionParameters::JsParameters(params);
    let body: AnyJsFunctionBody = AnyJsFunctionBody::JsFunctionBody(body);
    js_arrow_function_expression(params, token(T![=>]), body).build()
}

pub fn make_js_parameters(
    js_parameter_list: JsParameterList, // parameters: Vec<AnyJsParameter>,
                                        // separators: Vec<JsSyntaxToken>,
) -> JsParameters {
    js_parameters(token(T!['(']), js_parameter_list, token(T![')']))
}

pub fn make_members() -> JsStatementList {
    js_statement_list(vec![])
}

pub fn make_js_return_statement(any_js_expression: AnyJsExpression) -> JsReturnStatement {
    js_return_statement(token(T![return]))
        .with_argument(any_js_expression)
        .with_semicolon_token(token(T![;]))
        .build()
}

pub fn make_js_function_body(
    directives: JsDirectiveList,
    statements: JsStatementList,
) -> JsFunctionBody {
    js_function_body(token(T!['{']), directives, statements, token(T!['}']))
}

pub fn collect_jsx_element_tags(node: &JsxElement) {
    let opening_element = node
        .opening_element()
        .expect("JsxElement should have an opening element");
    let name = opening_element
        .name()
        .expect("Opening element should have a name");
    let attr = opening_element.attributes();
    let jsx_name = name.as_jsx_name().expect("Name should be a JsxName");
    let value_token = jsx_name
        .value_token()
        .expect("JsxName should have a value token");
    let name = value_token.text();
    let child_list = node.children();
    child_list.into_iter().for_each(|child| match child {
        AnyJsxChild::JsxElement(jsx_element) => {
            collect_jsx_element_tags(&jsx_element);
        }
        _ => {}
    });
}

pub fn collect_jsx_tag_expression(node: &JsxTagExpression) {
    let tag = node.tag().expect("JsxTagExpression should have a tag");
    let jsx_element = tag.as_jsx_element().expect("Tag should be a JsxElement");
    collect_jsx_element_tags(jsx_element);
}

// 用于还原 solidjs 构建语句的参数结构体
pub struct StatementItemConfig {
    pub el_var: String,                        // 变量名，如 "_el$"
    pub tmpl_fn: String,                       // 模板工厂函数名，如 "_tmpl$"
    pub event_bindings: Vec<(String, String)>, // 事件绑定，如 [("$$click", "increment")]
    pub inserts: Vec<(String, String)>,        // 插入操作，如 [("_el$", "count")]
    pub return_var: String,                    // 返回变量名，如 "_el$"
}

pub fn make_statement_items(config: &StatementItemConfig) -> Vec<AnyJsStatement> {
    let mut stmts = Vec::new();

    // 1. var _el$ = _tmpl$();
    let el_var_token = SyntaxToken::new_detached(T![ident], &config.el_var, Vec::new(), Vec::new());
    let tmpl_fn_token =
        SyntaxToken::new_detached(T![ident], &config.tmpl_fn, Vec::new(), Vec::new());
    let el_var = js_reference_identifier(el_var_token.clone());
    let tmpl_fn = js_reference_identifier(tmpl_fn_token.clone());
    let declarator = js_variable_declarator(AnyJsBindingPattern::AnyJsBinding(
        AnyJsBinding::JsIdentifierBinding(js_identifier_binding(el_var_token.clone())),
    ))
    .with_initializer(biome_js_syntax::JsInitializerClause::unwrap_cast(
        biome_js_factory::make::js_initializer_clause(
            token(T![=]),
            biome_js_syntax::AnyJsExpression::JsCallExpression(
                js_call_expression(
                    biome_js_syntax::AnyJsExpression::JsIdentifierExpression(
                        js_identifier_expression(tmpl_fn.clone()),
                    ),
                    js_call_arguments(
                        token(T!['(']),
                        js_call_argument_list(vec![], vec![]),
                        token(T![')']),
                    ),
                )
                .build(),
            ),
        )
        .syntax()
        .clone(),
    ))
    .build();
    let var_decl = js_variable_statement(
        js_variable_declaration(
            token(T![var]),
            js_variable_declarator_list(vec![declarator], vec![]),
        )
        .build(),
    )
    .build();
    stmts.push(biome_js_syntax::AnyJsStatement::JsVariableStatement(
        var_decl,
    ));

    // 2. 事件绑定: _el$.$$click = increment;
    for (event, handler) in &config.event_bindings {
        let event_token = SyntaxToken::new_detached(T![ident], event, Vec::new(), Vec::new());
        let handler_token = SyntaxToken::new_detached(T![ident], handler, Vec::new(), Vec::new());
        let member = js_static_member_assignment(
            biome_js_syntax::AnyJsExpression::JsIdentifierExpression(js_identifier_expression(
                el_var.clone(),
            )),
            token(T![.]),
            biome_js_syntax::AnyJsName::JsName(js_name(event_token)),
        );
        let assign = js_expression_statement(
            biome_js_syntax::AnyJsExpression::JsAssignmentExpression(js_assignment_expression(
                AnyJsAssignmentPattern::AnyJsAssignment(
                    biome_js_syntax::AnyJsAssignment::JsStaticMemberAssignment(member),
                ),
                token(T![=]),
                biome_js_syntax::AnyJsExpression::JsIdentifierExpression(js_identifier_expression(
                    js_reference_identifier(handler_token),
                )),
            )),
        )
        .build();
        stmts.push(biome_js_syntax::AnyJsStatement::JsExpressionStatement(
            assign,
        ));
    }
 
    // 3. 插入操作: _$insert(_el$, count);
    for (el, value) in &config.inserts {
        let el_token = SyntaxToken::new_detached(T![ident], el, Vec::new(), Vec::new());
        let value_token = SyntaxToken::new_detached(T![ident], value, Vec::new(), Vec::new());
        let insert_token = SyntaxToken::new_detached(T![ident], "_$insert", Vec::new(), Vec::new());
        let call = js_expression_statement(AnyJsExpression::JsCallExpression(
            js_call_expression(
                AnyJsExpression::JsIdentifierExpression(js_identifier_expression(
                    js_reference_identifier(insert_token),
                )),
                js_call_arguments(
                    token(T!['(']),
                    js_call_argument_list(
                        vec![
                            AnyJsCallArgument::AnyJsExpression(
                                AnyJsExpression::JsIdentifierExpression(js_identifier_expression(
                                    js_reference_identifier(el_token.clone()),
                                )),
                            ),
                            AnyJsCallArgument::AnyJsExpression(
                                AnyJsExpression::JsIdentifierExpression(js_identifier_expression(
                                    js_reference_identifier(value_token.clone()),
                                )),
                            ),
                        ],
                        vec![],
                    ),
                    token(T![')']),
                ),
            )
            .build(),
        ))
        .build();
        stmts.push(AnyJsStatement::JsExpressionStatement(call));
    }

    // 4. return _el$;
    let ret = make_js_return_statement(js_identifier_expression(el_var).into()).into();
    stmts.push(ret);

    stmts
}
