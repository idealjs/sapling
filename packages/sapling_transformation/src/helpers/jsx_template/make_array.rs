use biome_js_factory::make::{js_array_element_list, js_array_expression, token};
use biome_js_syntax::{AnyJsArrayElement, JsArrayExpression, T};

pub fn make_array(elements: Vec<AnyJsArrayElement>) -> JsArrayExpression {
    js_array_expression(
        token(T!['[']),
        js_array_element_list(
            elements.clone(),
            vec![token(T![,]); elements.len().saturating_sub(1)],
        ),
        token(T![']']),
    )
}
