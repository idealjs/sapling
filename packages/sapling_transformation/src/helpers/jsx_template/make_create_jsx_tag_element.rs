use biome_js_factory::make::{
    js_arrow_function_expression, js_call_argument_list, js_call_arguments, js_call_expression,
    js_directive_list, js_function_body, js_identifier_expression, js_parameter_list,
    js_parameters, js_reference_identifier, token,
};
use biome_js_syntax::{
    AnyJsExpression, AnyJsStatement, JsCallExpression, JsDirective, JsSyntaxToken, T,
};

pub fn make_create_jsx_tag_element(
    directives: Vec<JsDirective>,
    statements: Vec<AnyJsStatement>,
) -> JsCallExpression {
    let function_body = js_function_body(
        token(T!['{']),
        js_directive_list(directives),
        biome_js_factory::make::js_statement_list(statements),
        token(T!['}']),
    );
    let params = js_parameters(
        token(T!['(']),
        js_parameter_list(vec![], vec![]),
        token(T![')']),
    );
    let arrow_fn =
        js_arrow_function_expression(params.into(), token(T![=>]), function_body.into()).build();

    let create_jsx_tag_element_token =
        JsSyntaxToken::new_detached(T![ident], "_$createJsxTagElement", Vec::new(), Vec::new());

    js_call_expression(
        js_identifier_expression(js_reference_identifier(create_jsx_tag_element_token)).into(),
        js_call_arguments(
            token(T!['(']),
            js_call_argument_list(
                vec![biome_js_syntax::AnyJsCallArgument::AnyJsExpression(
                    AnyJsExpression::JsArrowFunctionExpression(arrow_fn),
                )],
                vec![],
            ),
            token(T![')']),
        ),
    )
    .build()
}
