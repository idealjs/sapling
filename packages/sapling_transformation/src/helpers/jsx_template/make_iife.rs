use biome_js_factory::make::{
    js_arrow_function_expression, js_call_argument_list, js_call_arguments, js_call_expression,
    js_function_body, js_parameter_list, js_parameters, js_parenthesized_expression, token,
};
use biome_js_syntax::{AnyJsExpression, JsCallExpression, JsDirectiveList, JsStatementList, T};

pub fn make_iife(directives: JsDirectiveList, statements: JsStatementList) -> JsCallExpression {
    let function_body = js_function_body(token(T!['{']), directives, statements, token(T!['}']));
    let params = js_parameters(
        token(T!['(']),
        js_parameter_list(vec![], vec![]),
        token(T![')']),
    );
    let arrow_fn =
        js_arrow_function_expression(params.into(), token(T![=>]), function_body.into()).build();
    js_call_expression(
        js_parenthesized_expression(
            token(T!['(']),
            AnyJsExpression::JsArrowFunctionExpression(arrow_fn),
            token(T![')']),
        )
        .into(),
        js_call_arguments(
            token(T!['(']),
            js_call_argument_list(vec![], vec![]),
            token(T![')']),
        ),
    )
    .build()
}
