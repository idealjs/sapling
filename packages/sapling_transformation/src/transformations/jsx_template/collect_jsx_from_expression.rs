use biome_js_syntax::*;
use crate::JsxElementInfo;

pub fn collect_jsx_from_expression(expr: &AnyJsExpression, jsx_elements: &mut Vec<JsxElementInfo>, in_function: bool) {
    match expr {
        AnyJsExpression::JsxTagExpression(jsx_tag) => {
            if let Ok(jsx_element_any) = jsx_tag.tag() {
                match jsx_element_any {
                    AnyJsxTag::JsxElement(element) => {
                        if let Ok(opening) = element.opening_element() {
                            if let Ok(name) = opening.name() {
                                if let Some(jsx_name) = name.as_jsx_name() {
                                    if let Ok(name_token) = jsx_name.value_token() {
                                        let tag_name = name_token.text_trimmed().to_string();
                                        jsx_elements.push(JsxElementInfo {
                                            tag_name,
                                            position: jsx_elements.len(),
                                            is_self_closing: false,
                                            in_function,
                                        });
                                    }
                                }
                            }
                        }
                    },
                    AnyJsxTag::JsxSelfClosingElement(self_closing) => {
                        if let Ok(name) = self_closing.name() {
                            if let Some(jsx_name) = name.as_jsx_name() {
                                if let Ok(name_token) = jsx_name.value_token() {
                                    let tag_name = name_token.text_trimmed().to_string();
                                    jsx_elements.push(JsxElementInfo {
                                        tag_name,
                                        position: jsx_elements.len(),
                                        is_self_closing: true,
                                        in_function,
                                    });
                                }
                            }
                        }
                    },
                    _ => {},
                }
            }
        },
        AnyJsExpression::JsArrowFunctionExpression(arrow_fn) => {
            if let Ok(body) = arrow_fn.body() {
                if let AnyJsFunctionBody::AnyJsExpression(expr) = body {
                    collect_jsx_from_expression(&expr, jsx_elements, true);
                }
            }
        },
        AnyJsExpression::JsParenthesizedExpression(paren_expr) => {
            if let Ok(inner_expr) = paren_expr.expression() {
                collect_jsx_from_expression(&inner_expr, jsx_elements, in_function);
            }
        },
        _ => {},
    }
}