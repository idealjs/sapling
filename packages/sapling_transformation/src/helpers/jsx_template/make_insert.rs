// 用于生成 _$insert(parent_id, expr) 的辅助函数
//
// # 参数
// - parent_id: 父节点标识符（字符串）
// - expr: 需要插入的表达式（AnyJsExpression）
//
// # 返回
// 返回 JsCallExpression，表示 _$insert(parent_id, expr) 的调用。
// 用于 JSX 转换流程中插入子节点或表达式。
//
// # 用法示例
// let call_expr = make_insert("_el$", some_expr);

use crate::helpers::jsx_template::make_js_call_arguments;
use biome_js_factory::make::{
    js_call_expression, js_identifier_expression, js_reference_identifier, token,
};
use biome_js_syntax::{AnyJsCallArgument, AnyJsExpression, JsCallExpression, T};

/// 生成 _$insert(parent_id, expr) 的 JsCallExpression
pub fn make_insert(parent_id: &str, expr: AnyJsExpression) -> JsCallExpression {
    let callee = js_identifier_expression(js_reference_identifier(biome_js_factory::make::ident(
        "_$insert",
    )));
    let arg1 = AnyJsCallArgument::AnyJsExpression(AnyJsExpression::JsIdentifierExpression(
        js_identifier_expression(js_reference_identifier(biome_js_factory::make::ident(
            parent_id,
        ))),
    ));
    let arg2 = AnyJsCallArgument::AnyJsExpression(expr);
    js_call_expression(
        callee.into(),
        make_js_call_arguments(vec![arg1, arg2], vec![token(T!(,))]),
    )
    .build()
}
