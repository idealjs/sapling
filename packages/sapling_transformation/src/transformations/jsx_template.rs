// use crate::{JsBatchMutation, declare_transformation};
use biome_analyze::context::RuleContext;
use biome_analyze::{Ast, Rule};
use biome_js_factory::make::{
    ident, js_assignment_expression, js_call_argument_list, js_call_arguments, js_call_expression,
    js_computed_member_assignment, js_decorator_list, js_directive_list, js_expression_statement,
    js_formal_parameter, js_function_body, js_function_expression, js_identifier_assignment,
    js_identifier_binding, js_identifier_expression, js_logical_expression, js_module_item_list,
    js_number_literal_expression, js_object_expression, js_object_member_list, js_parameter_list,
    js_parameters, js_parenthesized_expression, js_reference_identifier, js_statement_list,
    js_string_literal, js_string_literal_expression, js_variable_declaration,
    js_variable_declarator, js_variable_declarator_list, js_variable_statement, token,
};
use biome_js_syntax::{
    AnyJsAssignment, AnyJsAssignmentPattern, AnyJsBinding, AnyJsBindingPattern, AnyJsCallArgument,
    AnyJsExpression, AnyJsFormalParameter, AnyJsLiteralExpression, AnyJsModuleItem, AnyJsParameter,
    AnyJsStatement, JsAssignmentExpression, JsComputedMemberAssignment, JsExpressionStatement,
    JsFunctionExpression, JsInitializerClause, JsLogicalExpression, JsModuleItemList,
    JsReturnStatement, JsStatementList, JsSyntaxKind, JsVariableStatement, JsxElement,
    JsxTagExpression, T, TsEnumDeclaration,
};
use biome_rowan::{AstNode, BatchMutationExt, TriviaPieceKind};

use crate::helpers::jsx_template::{
    StatementItemConfig, collect_jsx_tag_expression, make_js_arrow_function_expression,
    make_js_call_expression, make_js_function_body, make_js_parameters, make_js_return_statement,
    make_statement_items,
};
use crate::{JsBatchMutation, declare_transformation};

declare_transformation! {
    /// Transform a TypeScript [TsEnumDeclaration]
    pub(crate) JsxTemplate {
        version: "0.1.0",
        name: "jsx_template",
        language: "js",
    }
}

#[derive(Debug)]
pub struct JsxTemplateMembers {
    name: String,
    attr_names: Vec<(String, Option<JsInitializerClause>)>,
}

impl Rule for JsxTemplate {
    type Query = Ast<JsxTagExpression>;
    type State = ();
    type Signals = Option<Self::State>;
    type Options = ();

    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        let node = ctx.query();
        // let child_list = node.children();
        // let attrs = node
        //     .opening_element()
        //     .expect("JsxElement should have an opening element")
        //     .attributes();

        println!("node run : {:?}", node);
        collect_jsx_tag_expression(node);
        // let mut member_names = vec![];
        // let id = node.id().ok()?;
        Some(())
    }

    fn transform(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsBatchMutation> {
        let node = ctx.query();
        let mut mutation = node.clone().begin();
        let parent = node.syntax().parent();
        println!("node transform : {:?}", node);

        if let Some(parent) = parent {
            match parent.kind() {
                JsSyntaxKind::JS_RETURN_STATEMENT => {
                    if let Some(prev_node) = JsReturnStatement::cast(parent) {
                        let next_node =
                            make_js_return_statement(AnyJsExpression::JsCallExpression(
                                make_js_call_expression(make_js_arrow_function_expression(
                                    make_js_parameters(js_parameter_list(vec![], vec![])),
                                    make_js_function_body(
                                        js_directive_list(vec![]),
                                        js_statement_list(make_statement_items(
                                            &StatementItemConfig {
                                                el_var: "_el$".to_string(),
                                                tmpl_fn: "_tmpl$".to_string(),
                                                event_bindings: vec![(
                                                    "$$click".to_string(),
                                                    "increment".to_string(),
                                                )],
                                                inserts: vec![(
                                                    "_el$".to_string(),
                                                    "count".to_string(),
                                                )],
                                                return_var: "_el$".to_string(),
                                            },
                                        )),
                                    ),
                                )),
                            ));

                        mutation.replace_node(prev_node, next_node);
                    }
                }
                _ => {}
            }
        }
        // None
        Some(mutation)
    }
}
