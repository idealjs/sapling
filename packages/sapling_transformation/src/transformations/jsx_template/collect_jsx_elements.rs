use biome_rowan::AstNodeList;
use biome_js_syntax::*;
use crate::JsxElementInfo;
use crate::jsx_template::{collect_jsx_from_statement, collect_jsx_from_expression};

pub fn collect_jsx_elements(item: &AnyJsModuleItem, jsx_elements: &mut Vec<JsxElementInfo>) {
    match item {
        AnyJsModuleItem::AnyJsStatement(stmt) => {
            collect_jsx_from_statement(stmt, jsx_elements, false);

            // 递归处理 function 声明体内的语句
            if let AnyJsStatement::JsFunctionDeclaration(func_decl) = stmt {
                if let Ok(body) = func_decl.body() {
                    let stmts = body.statements();
                    for inner_stmt in stmts.iter() {
                        collect_jsx_from_statement(&inner_stmt, jsx_elements, true);
                    }
                }
            }
        },
        AnyJsModuleItem::JsExport(export) => {
            // 简化版本：仅处理导出的表达式
            if let Ok(clause) = export.export_clause() {
                if let AnyJsExportClause::JsExportDefaultDeclarationClause(decl) = clause {
                    if let Ok(declaration) = decl.declaration() {
                        if let AnyJsExportDefaultDeclaration::JsFunctionExportDefaultDeclaration(expr_decl) = declaration {
                            if let Ok(body) = expr_decl.body() {
                                let list = body.statements();
                                for stmt in list.iter() {
                                    collect_jsx_from_statement(&stmt, jsx_elements, true);
                                }
                            }
                        }
                    }
                }
            }
        },
        _ => {},
    }
}