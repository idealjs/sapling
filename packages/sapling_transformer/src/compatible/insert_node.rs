use biome_js_factory::make::{
    ident, js_call_expression, js_identifier_expression, js_reference_identifier, token,
};
use biome_js_syntax::{AnyJsCallArgument, AnyJsExpression, JsCallExpression, T};
use sapling_transformation::helpers::jsx_template::make_js_call_arguments;

/// 生成 _$insertNode(parent_id, child_id) 的表达式语句
pub fn generate_insert_node_expr(parent_id: &str, child_id: &str) -> JsCallExpression {
    let callee = js_identifier_expression(js_reference_identifier(ident("_$insertNode")));
    let arg1 = AnyJsCallArgument::AnyJsExpression(AnyJsExpression::JsIdentifierExpression(
        js_identifier_expression(js_reference_identifier(ident(parent_id))),
    ));
    let arg2 = AnyJsCallArgument::AnyJsExpression(AnyJsExpression::JsIdentifierExpression(
        js_identifier_expression(js_reference_identifier(ident(child_id))),
    ));
    js_call_expression(
        callee.into(),
        make_js_call_arguments(vec![arg1, arg2], vec![token(T!(,))]),
    )
    .build()
}

#[test]
fn test_generate_insert_node_expr() {
    let stmt = generate_insert_node_expr("_el$", "_child$");
    // 只做快照测试，确保语法结构正确
    insta::assert_snapshot!(stmt.to_string());
}
