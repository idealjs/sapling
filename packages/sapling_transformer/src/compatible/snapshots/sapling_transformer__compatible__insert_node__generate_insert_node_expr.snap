---
source: packages/sapling_transformer/src/compatible/insert_node.rs
expression: stmt.to_string()
---
_$insertNode(_el$,_child$)
